Scala 3 migration

MateuszKubuszok opened the pull request:
TODO:

 - [x] remaining `implicit` -> `given`
 - [x] replace Monocle with Quicklens
 - [ ] make sure every `.query[A]` becomes `.queryWithLabel[A](label)`, same with `.update` and `.updateWithLabel(label)`
 - [ ] fix MDC (implement "stack")
 - [ ] test new Jsoniter implementation for OpenAPI (they changed something again...)
 - [ ] fix all new warnings
 - [ ] restore `Show.wrap`
 - [ ] fix GHA and migrate from Makefile to Justfile

MateuszKubuszok committed:
Update dependencies for Scala 3 migration (replace newtype+refined with neotype, remove compiler plugins)

MateuszKubuszok committed:
Migrated commonJVM module

MateuszKubuszok committed:
Module common-api compiles with Scala 3

MateuszKubuszok committed:
Make common-infrastructure compile

MateuszKubuszok committed:
Make modules discussions, discussionsApi, users, usersApi compile, redo sbt config

MateuszKubuszok committed:
Make discussionsImpl and usersImpl compile on Scala 3

MateuszKubuszok committed:
Make discussionsImpl and usersImpl tests compile

MateuszKubuszok committed:
Make server module compile

MateuszKubuszok committed:
Almost compiling app module

MateuszKubuszok committed:
Make app compile

MateuszKubuszok committed:
Make server module and its tests compile

MateuszKubuszok committed:
Replace all implicits (where possible) with given/using/extension

MateuszKubuszok committed:
Fix most warnings

MateuszKubuszok committed:
Make jFairy runnable at all, create Justfile

xiaobaozi0 commented:
@MateuszKubuszok  About OpenAPI, the default version is 3.1.0 but the swagger UI supports only 3.0.x. A quick fix is to use `.openapi("3.0.3")`.
```scala
def openAPI: OpenAPI = OpenAPIDocsInterpreter(OpenAPIServer.openAPIDocsOptions)
    .toOpenAPI(endpoints.map(_.endpoint).toList, apiInfo.toOpenAPI)
    .openapi("3.0.3")
    // TODO: quicklens
    .pipe(oa =>
      oa.copy(paths = oa.paths.copy(pathItems = oa.paths.pathItems.view.mapValues(fixPathItem).to(ListMap))
      )
    )
    ```

xiaobaozi0 commented on modules/app/src/main/scala/io/branchtalk/Program.scala:
@MateuszKubuszok Should be usersConfig ?

MateuszKubuszok committed:
Finish Scala 3.8.4 migration: swap to Kindlings/Hearth, fix Updatable avro bug

- Scala 3.4.2 -> 3.8.4; sbt/plugins bumped; drop Scala.js axis (JVM-only)
- Kindlings/Hearth replace: avro4s -> kindlings-avro, custom jsoniter -> kindlings-jsoniter,
  typesafe config+pureconfig -> sconfig+kindlings-sconfig (ekrich), sourcecode -> hearth.source,
  Kittens+Magnolia FastEq/ShowPretty -> kindlings-cats-derivation + kindlings-fast-show-pretty,
  custom OpenAPI jsoniter -> kindlings-tapir-openapi-jsoniter; add kindlings-cats-integration
- Fix long-standing IT failure: Kindlings' structural avro derivation of the generic Updatable
  enum names the Set union branch inconsistently (encoder Set__Type vs decoder Set), silently
  dropping update events. Provide explicit AvroCodec[Updatable[A]] as union[null,A]; give
  Post.Content a concrete codec. Unsuppress io.branchtalk test loggers (error) that hid it.
- docker-compose: Bitnami images gone -> postgres:12-alpine, apache/kafka:3.7.0 (KRaft), redis:7
- CI: modernize actions (checkout@v4, setup-java/temurin-17, setup-sbt), Makefile docker compose v2

MateuszKubuszok committed:
Format derivation.scala (scalafmt align)

MateuszKubuszok committed:
Replace per-companion newtype avro givens with a local neotype IsValueType provider

Bump Kindlings 0.3.0 -> 0.3.1 and Hearth 0.4.0 -> 0.4.1, and finish the neotype
avro story locally (no Kindlings integration module).

- Drop the hand-written AvroCodec[Updatable[A]] and Post.Content codecs: the
  generic-enum record-naming bug they worked around is fixed in Kindlings 0.3.1
  (verified via standalone repros); the enclosing event records derive them
  structurally now.

- Add modules/neotype-kindlings: a Hearth StandardMacroExtension registering an
  IsValueType provider for all four neotype flavours used here (neotype
  Newtype/Subtype + branchtalk's higher-kinded NewtypeT/NewtypeF). It unwraps a
  neotype opaque type to its underlying (translucent super type) and re-wraps by
  identity, matching structurally at macro-expansion time. This sidesteps the
  transparent-inline `Newtype.WithType` given that macro-time Expr.summon cannot
  see, which is why every neotype needed a concrete per-companion
  `given AvroCodec[X] = AvroSupport.newtypeCodec`. Loaded via ServiceLoader; wired
  onto the macro classpath by making `common` depend on the module.

- Remove all 32 per-companion `given AvroCodec[X] = AvroSupport.newtypeCodec` and
  the now-dead newtypeCodec helper + generic given.

- Move the base URI/UUID/OffsetDateTime/Array[Byte] codecs into AvroSupport so the
  `import AvroSupport.{ *, given }` already present at every avro-derivation site
  brings them into scope (newtypes are now unwrapped to their underlying at the
  derivation site, and `import shared.model.*` does not import givens under
  -source 3.3-migration).

Note: the provider's identity re-wrap skips neotype `validate` on the avro decode
path (the old newtypeCodec ran `make`); self-produced round-trips are unaffected.

MateuszKubuszok committed:
Fix CI: flatten encrypted user-event subtypes and make base codecs consistent

Two issues surfaced by CI on the previous commit:

1. Pre-existing Kindlings sum-derivation bug (fails on 0.3.0 and 0.3.1, independent
   of the neotype provider): a sealed subtype nested as `Create.Encrypted` inside
   the companion object of a same-named sibling case class (`Create`) makes the
   derived avro sum encoder MatchError on that subtype at runtime. Flatten the
   encrypted variants of UserEvent (Created/Updated) and UserCommandEvent
   (Create/Update) to top-level subtypes `CreatedEncrypted`/`UpdatedEncrypted` /
   `CreateEncrypted`/`UpdateEncrypted`, and rename usages in the writes/projectors.

2. My regression from the provider change: the neotype provider unwraps a newtype
   to its underlying at each derivation site, so branchtalk's base-type codecs
   (OffsetDateTime-as-String etc., which deliberately differ from Kindlings'
   built-in long/uuid-logical) must be in scope there. Kindlings inline re-derives
   nested types, so `UsersCommandEvent` (which wraps UserCommandEvent) needs the
   base givens too - it was the only avro-deriving site missing
   `import AvroSupport.{ *, given }`. Without it the schema used `long` while the
   encoder wrote a String -> ClassCastException. Import precedence makes
   branchtalk's codec win over Kindlings' built-in where imported (verified).
